mempool/mempool.go

```diff
@@ -12,70 +13,137 @@ import (
-	// done. E.g. from CheckTx.
-	// NOTE: Lock/Unlock must be managed by caller
-	FlushAppConn() error
-
-	// Flush removes all transactions from the mempool and cache
-	Flush()
-
-	// TxsAvailable returns a channel which fires once for every height,
-	// and only when transactions are available in the mempool.
-	// NOTE: the returned channel may be nil if EnableTxsAvailable was not called.
-	TxsAvailable() <-chan struct{}
-
-	// EnableTxsAvailable initializes the TxsAvailable channel, ensuring it will
-	// trigger once every height when transactions are available.
-	EnableTxsAvailable()
-
-	// Size returns the number of transactions in the mempool.
-	Size() int
-
-	// TxsBytes returns the total size of all txs in the mempool.
-	TxsBytes() int64
-
-	// InitWAL creates a directory for the WAL file and opens a file itself. If
-	// there is an error, it will be of type *PathError.
-	InitWAL() error
-
-	// CloseWAL closes and discards the underlying WAL file.
-	// Any further writes will not be relayed to disk.
-	CloseWAL()
+type (
+	Mempool[T TXReaper] interface {
+		// CheckTx executes a new transaction against the application to determine
+		// its validity and whether it should be added to the mempool.
+		CheckTx(tx types.Tx, callback func(*abci.Response), txInfo TxInfo) error
+
+		// PrepBlockFinality prepares the mempool for finalizing a block. This may require
+		// locking or other concerns that must be handled before a block is safe to commit.
+		PrepBlockFinality(ctx context.Context) (finishFn func(), err error)
+
+		// Reap allows the caller to reap certificate digests used for proposal. Callers
+		// can choose whatever predicate suits their needs. All predicates are ANDed, aka
+		// all set predicates must be satisfied. If no options are provided, then all TXs
+		// will be returned.
+		Reap(ctx context.Context, opts ...ReapOptFn) (T, error)
+
+		// Update informs the mempool that the given txs were committed and can be discarded.
+		// NOTE: this should be called *after* block is committed by consensus.
+		// NOTE: Lock/Unlock must be managed by caller
+		Update(
+			blockHeight int64,
+			blockTxs types.Txs,
+			deliverTxResponses []*abci.ResponseDeliverTx,
+			newPreFn PreCheckFunc,
+			newPostFn PostCheckFunc,
+		) error
+
+		// Flush removes all transactions from the mempool and cache
+		Flush()
+
+		// TxsAvailable returns a channel which fires once for every height,
+		// and only when transactions are available in the mempool.
+		// NOTE: the returned channel may be nil if EnableTxsAvailable was not called.
+		TxsAvailable() <-chan struct{}
+
+		// EnableTxsAvailable initializes the TxsAvailable channel, ensuring it will
+		// trigger once every height when transactions are available.
+		EnableTxsAvailable()
+
+		// Meta returns metadat for the mempool.
+		Meta() Meta
+
+		// InitWAL creates a directory for the WAL file and opens a file itself. If
+		// there is an error, it will be of type *PathError.
+		InitWAL() error
```

Comment:
all the WAL stuff can be bottled up too, this feels like something that doesn't belong in the actual mempool interface. My thinking is the mempool would have its WAL inited as needed. Feels odd to expose these implementation details 🤔